loading.bs

```diff
@@ -0,0 +1,949 @@
+[=byte sequence=] |headerBytes| is the concatenation of the following [=byte
+sequences=]:
+
+1. The byte 0x20 (SP) repeated 64 times. This matches the TLS 1.3 ([[RFC8446]])
+    format to avoid cross- protocol attacks if anyone uses the same key in a TLS
+    certificate and an exchange-signing certificate.
+1. `` `HTTP Exchange 1 b1` `` to serve as a context string.
+
+    Note: Each draft of
+    [[draft-yasskin-httpbis-origin-signed-exchanges-impl-01]] and the final RFC
+    for [[draft-yasskin-http-origin-signed-responses]] will use distinct context
+    strings.
+1. A single 0x00 byte which serves as a separator.
+1. The concatenation of the following [=byte sequences=], which is the
+    [=canonically-encoded CBOR=] for a CBOR map:
+    1. 0xA5, the CBOR header for a 5-item map.
+    1. 0x64 followed by `` `date` ``
+    1. The [=CBOR integer encoding=] of |signature|'s [=exchange
+        signature/date=].
+    1. 0x67 followed by `` `expires` ``.
+    1. The [=CBOR integer encoding=] of |signature|'s [=exchange
+        signature/expiration time=].
+    1. 0x67 followed by `` `headers` ``.
+    1. |headerBytes|.
+    1. 0x6B followed by `` `cert-sha256` ``.
+    1. The [=CBOR byte string encoding=] of |signature|'s [=exchange
+        signature/certSha256=].
+    1. 0x6C followed by `` `validity-url` ``.
+    1. The [=CBOR byte string encoding=] of |signature|'s [=exchange
+        signature/validityUrlBytes=].
+
+<h3 algorithm id="validating-signature">Validating a signature</h3>
+
+An [=exchange signature=] |signature| <dfn for="exchange signature" lt="is
+valid|is not valid">is valid</dfn> for a [=byte sequence=] |headerBytes| if the
+following steps return `valid`:
+
+1. Let |clockSkew| be the uncertainty in the UA's estimate of the current time
+    caused by clock skew on the client. The UA MAY set this to 0 or use a more
+    sophisticated estimate.
+1. If the UA's estimate of the current time is more than |clockSkew| before
+    |signature|'s [=exchange signature/date=], return "untrusted".
+
+    Note: We take estimated clock skew into account when checking the
+    signature's [=exchange signature/date=] because we want well-behaved servers
+    to use the time they created the signature, but if they immediately start
+    serving that signature, and skewed clients don't try to correct for their
+    skew, those clients will reject the signature.
+1. If the UA's estimate of the current time is after |signature|'s [=exchange
+    signature/expiration time=], return "untrusted".
+
+    Note: We use the client's best guess of the current time to check the
+    [=exchange signature/expiration time=] so that attackers trying to get an
+    exchange trusted for longer, are constrained to modify the client's clock
+    and can't also attack its estimate of its skew.
+1. Let |message| be the [=b1 signed message=] for |signature| and |headerBytes|.
+1. Let |publicKey| be the [=certificate/public key=] of |parsedSignature|'s
```

Comment:
@sleevi @davidben The fact that the [certificate](http://api.csswg.org/bikeshed/?url=https://raw.githubusercontent.com/jyasskin/webpackage/loading-spec/loading.bs#augmented-certificate-certificate) is stored un-parsed means that the spec winds up saying to parse out a public key twice, which feels like a security risk. How much can systems actually parse these structures ahead of time, vs when they're validating origin-trust?

Reply:
Can you explain why it feels like a security risk? I'm not sure I fully understand the question. Extracting an SPKI from a certificate is something easy on virtually all systems - the question is whether or not it's "trusted" data yet.

Reply:
I presume they're going to be using the same parser - is there reason to believe they may be different?

Reply:
No particular reason.